lib/charms/nrpe_exporter/v0/nrpe_exporter.py

```diff
@@ -0,0 +1,342 @@
+
+
+class NrpeEvents(ObjectEvents):
+    """Event descriptor for events raised by `NrpeExporterProvider`."""
+
+    nrpe_targets_changed = EventSource(NrpeTargetsChangedEvent)
+
+
+class NrpeExporterProvider(Object):
+    """A NRPE exporter based monitor."""
+
+    on = NrpeEvents()
+
+    def __init__(self, charm: CharmBase, relation_names: Optional[dict] = None):
+        """A NRPE exporter based monitor.
+
+        Args:
+            charm: a `CharmBase` instance that manages this
+                instance of the NRPE Exporter.
+            relation_names: an optional `dict` containing the interfaces to monitor.
+
+        Raises:
+            RelationNotFoundError: If there is no relation in the charm's metadata.yaml
+                with the same name as provided via `relation_name` argument.
+            RelationInterfaceMismatchError: The relation with the same name as provided
+                via `relation_name` argument does not have the `prometheus_scrape` relation
+                interface.
+            RelationRoleMismatchError: If the relation with the same name as provided
+                via `relation_name` argument does not have the `RelationRole.requires`
+                role.
+        """
+        relation_names = relation_names or DEFAULT_RELATION_NAMES
+        for relation_name, relation_interface in relation_names.items():
+            _validate_relation_by_interface_and_direction(
+                charm, relation_name, relation_interface, RelationRole.requires
+            )
+
+        super().__init__(charm, "_".join(relation_names.keys()))
+        self._charm = charm
+        self._relation_names = relation_names
+        for relation_name in relation_names.keys():
+            events = self._charm.on[relation_name]
+            self.framework.observe(events.relation_changed, self._on_nrpe_relation_changed)
+            self.framework.observe(events.relation_departed, self._on_nrpe_relation_changed)
+
+    def _on_nrpe_relation_changed(self, event: RelationEvent):
+        """Handle changes with related endpoints.
+
+        Anytime there are changes in relations between the NRPE exporter
+        and NRPE endpoints, the relation tree is scraped, and configurations
+        are generated, which can be watched through a `NrpeTargetsChangedEvent`.
+
+        Args:
+            event: a `RelationEvent` signifying a change.
+        """
+        rel_id = event.relation.id
+
+        self.on.nrpe_targets_changed.emit(relation_id=rel_id)
+
+    def endpoints(self) -> list:
+        """Fetch the list of NRPE exporter targets.
+
+        Returns:
+            A list consisting of all the endpoints and partial configurations
+            to be ingested by Prometheus.
+        """
+        nrpe_endpoints = []
+
+        for relation_name in self._relation_names.keys():
+            for relation in self._charm.model.relations[relation_name]:
+                nrpe_endpoints.extend(self._nrpe_scrape_config(relation))
```

Comment:
Is uniqueness a concern?

Reply:
Not at all, since these are per-unit in any case